tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala

```diff
@@ -0,0 +1,139 @@
+package at.forsyte.apalache.tla.typecmp.raw
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.oper.{TlaOper, TlaSetOper}
+import at.forsyte.apalache.tla.lir.values.TlaStr
+import at.forsyte.apalache.tla.typecmp.BuilderUtil.throwMsg
+import at.forsyte.apalache.tla.typecmp.{BuilderTypeException, BuilderUtil, TypeComputation, TypeComputationReturn}
+
+import scala.collection.immutable.SortedMap
+
+/**
+ * Raw builder for TlaSetOper expressions.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+trait RawSetBuilder extends ProtoBuilder {
+
+  /** {x1, ..., xn}, must have >0 args. */
+  protected def _enumSet(arg: TlaEx, args: TlaEx*): TlaEx = simpleInstruction(TlaSetOper.enumSet, arg +: args: _*)
+
+  /** {} : Set(elemType) */
+  protected def _emptySet(elemType: TlaType1): TlaEx = OperEx(TlaSetOper.enumSet)(Typed(SetT1(elemType)))
+
+  /** elem \in set */
+  protected def _in(elem: TlaEx, set: TlaEx): TlaEx = simpleInstruction(TlaSetOper.in, elem, set)
+
+  /** elem \notin set */
+  protected def _notin(elem: TlaEx, set: TlaEx): TlaEx = simpleInstruction(TlaSetOper.notin, elem, set)
+
+  /** left \cap right, left \intersect right */
+  protected def _cap(left: TlaEx, right: TlaEx): TlaEx = simpleInstruction(TlaSetOper.cap, left, right)
+
+  /** left \cup right, left `\`union right */
+  protected def _cup(left: TlaEx, right: TlaEx): TlaEx = simpleInstruction(TlaSetOper.cup, left, right)
+
+  /** UNION set */
+  protected def _union(set: TlaEx): TlaEx = simpleInstruction(TlaSetOper.union, set)
+
+  /** { x \in set: p } */
+  protected def _filter(x: NameEx, set: TlaEx, p: TlaEx): TlaEx =
+    simpleInstruction(TlaSetOper.filter, x, set, p)
+
+  /** { e: x1 \in set1 , ..., xN \in setN }, must have at least 1 var-set pair */
+  protected def _map(
+      e: TlaEx,
+      x1: NameEx,
+      set1: TlaEx,
+      pairs: TlaEx*): TlaEx = {
```

Comment:
I found it annoying that the builder differentiates between the first parameter and the other parameters. So I would prefer just `TlaEx*` or maybe `(NameEx, TlaEx)*`, which you can nicely write as `n1 -> e1, n2 -> e2, ...`. Perhaps, shapeless can help us in doing static checks to make sure that the list is non-empty. @shonfeder ?